src/nvim/normal.c

```diff
@@ -1236,8 +1236,9 @@ static void normal_redraw(NormalState *s)
 
   // If the cursor moves horizontally when 'concealcursor' is active, then the
   // current line needs to be redrawn in order to calculate the correct
-  // cursor position.
+  // cursor position.  The last line needs to be redrawn to prevent artifacts.
   if (curwin->w_p_cole > 0 && conceal_cursor_line(curwin)) {
+    redrawWinline(curwin, curwin->w_last_cursorline);
```

Comment:
This had nothing to do with horizontal movement as far as I can see. Redraw of previous cursorline is generally handled in `redraw_for_cursorline()`. Probably the check for `VALID` there is too sharp, we also need this in `INVERTED` and `INVERTED_ALL` for visual mode.

Reply:
Does it reproduce for you with the instructions in an `xterm`? I would help to debug the missing condition, but I cannot see the glitch to start with...

Reply:
`win_cursorline_standout` is `false` here - adding `VIsual_active ||` to it fixes it, but might be overkill?

Reply:
@bfredl 
Yes, also in xterm: `build/bin/nvim --noplugin -u t/11024.init.vim`, with:
```vim
help
set concealcursor=ncv
set conceallevel=2
```
Then `V` and holding j, then k.

Reply:
> win_cursorline_standout is false here - adding VIsual_active || to it fixes it, but might be overkill?

Should be handled by `conceal_cursor_line()` already with `get_real_state() & VISUAL`. We should figure out why it doesn't work, instead of papering over it.

Reply:
Though it actually disabled there, because it shouldn't be needed. We can of course force it to be true always (i e disregard `cocu+=v` in this check), but I would try a bit more to find the root cause of the glitch first (the glitchy state is not a valid line state to start with so it should not be drawn before it is overdrawn).

Reply:
Sure!
Added a test, hopefully that helps you with debugging it.

Reply:
It's also noteworthy that it never happens when doing it slowly, and also not when not triggering the window to scroll.

Reply:
Meant to post this before already: https://asciinema.org/a/J1vI0hvXsQFQiHXYszIrCpAyB

Reply:
@bfredl 
Could you look into it already?

Reply:
I did not know we are are in a hurry, but in case we are, just use the workaround I suggested for now.  i e inside `win_cursorline_standout` check visual mode before `conceal_cursor_line()` (and add a `TODO(bfredl)` about the underlying bug).

Reply:
No hurry - but there were two issues reported about it already.
It would be nice to fix this in 0.4.3, but I'd rather wait for you to fix it properly, of course.

Reply:
Taken it out into https://github.com/neovim/neovim/pull/11120 for now to keep it separate.